### ISSUE TITLE
Can't create logs when using managed package

### ISSUE THREAD

**Issue by Limbo20000** [jongpie/NebulaLogger#656]
### Package Edition of Nebula Logger

Managed Package

### Package Version of Nebula Logger

4.13

### New Bug Summary

We have tried a simple LWC component: 
```javascript
import { LightningElement } from 'lwc';

export default class ManagedPackageTest extends LightningElement {

  handleClick() {
    debugger;
    const logger = this.template.querySelector('Nebula-logger');
    logger.info('Hello World 🙋‍♀️');
    logger.saveLog();
  }
}
```
```html
<template>
  <Nebula-logger></Nebula-logger>
  <lightning-button variant="base" label="test" onclick={handleClick} ></lightning-button>
</template>
```
We have also enabled the LWS in session setup

**Comment by jongpie** [comment]
Hi @Limbo20000 - that approach is using the `logger` LWC in your markup, which is no longer the recommended approach. Can you please try using this approach instead ([shown in the `4.13.0` release notes](https://github.com/jongpie/NebulaLogger/releases/tag/v4.13.0))? This only requires changes to your JS file - the component is not needed in your HTML file.

```javascript
import { createLogger } from 'Nebula/logger';

export default class LoggerLWCDemo extends LightningElement {
  logger;
  
  async connectedCallback() {
    this.logger = await createLogger();
    this.logger.info('Hello, world');
    this.logger.saveLog();
  }
}
```

**Comment by Limbo20000** [comment]
Hi, we have tried this approach and the same issue persists. We have tried to console.log the logger after it should be created but it is undefined. The issue might be related to that it seems like the Nebula/logger namespace does not exists.

**Comment by jongpie** [comment]
@Limbo20000 thanks for trying that approach! Offhand, I'm not sure what the issue would be, but I'll try to do some testing to see if I can reproduce the issue & figure out what's going on.

**Comment by szymon-halik** [comment]
Hey @jongpie,

I work together with @Limbo20000 in the team. I've started to look into this issue today as well. The current configuration uses `EVENT_BUS` as save method. It's seems that `Nebula__LogEntryEvent__e` are received but the Log Entries are not created in our dev orgs(the Last Processed Id increases). I tried different save method but non of them creates Log Entries or Logs

I've created a simple dummy PE to check if events are received at all in our dev orgs:

```java
test__e testEvent = new test__e();
EventBus.publish(new List<test__e>{testEvent});
Nebula.Logger.error('some error');
Nebula.Logger.saveLog();
```   
![Screenshot 2024-03-22 at 11 17 59](https://github.com/jongpie/NebulaLogger/assets/35228022/150b9fb9-4b93-4028-844b-35780ed9fa77)
![Screenshot 2024-03-22 at 11 23 03](https://github.com/jongpie/NebulaLogger/assets/35228022/07f4a4a0-de56-4c81-823f-3c9d7b3a3866)

I would expect that `Nebula__LogEntryEvent__e` will publish and event but this is not happening. I tried to compile all classes in the org but without any luck.

Summing up, none of the save methods allows us to successfully persist records in the db. We are using the cloning technique for our dev orgs. The issue does not occur in any sandbox that is refreshed from prod

**Comment by jongpie** [comment]
Hi @szymon-halik - originally, I thought this was an issue with the LWC, but since your anonymous Apex script isn't working either, there has to be something else causing the problem. It's possibly an issue with how the `LoggerSettings__c` records have been configured in your org.

Could you run the Apex script below & share the JSON results from the debug log? This should provide enough details to know if the `LoggerSettings__c` data is the source of the problem.

```apex
System.debug(
  JSON.serializePretty(
    [
      SELECT
        Nebula__DefaultSaveMethod__c,
        Nebula__DefaultPlatformEventStorageLocation__c,
        Nebula__DefaultPlatformEventStorageLoggingLevel__c,
        Nebula__DefaultScenario__c,
        Nebula__EndTime__c,
        Nebula__IsApexSystemDebugLoggingEnabled__c,
        Nebula__IsJavaScriptConsoleLoggingEnabled__c,
        Nebula__IsEnabled__c,
        Nebula__LoggingLevel__c,
        Nebula__StartTime__c,
        SetupOwner.Type,
        SetupOwnerId
      FROM Nebula__LoggerSettings__c
    ]
  )
);
```

**Comment by szymon-halik** [comment]
@jongpie Thanks for the heads up. Here is the result of the above query:

```JSON
[ {
  "attributes" : {
    "type" : "Nebula__LoggerSettings__c",
    "url" : "/services/data/v60.0/sobjects/Nebula__LoggerSettings__c/a5K08000000kJtqEAE"
  },
  "Nebula__DefaultSaveMethod__c" : "EVENT_BUS",
  "Nebula__DefaultPlatformEventStorageLocation__c" : "CUSTOM_OBJECTS",
  "Nebula__IsApexSystemDebugLoggingEnabled__c" : false,
  "Nebula__IsJavaScriptConsoleLoggingEnabled__c" : false,
  "Nebula__IsEnabled__c" : true,
  "Nebula__LoggingLevel__c" : "ERROR",
  "SetupOwnerId" : "00DbW0000000JDxUAM",
  "Id" : "a5K08000000kJtqEAE",
  "SetupOwner" : {
    "attributes" : {
      "type" : "Name",
      "url" : "/services/data/v60.0/sobjects/Organization/00DbW0000000JDxUAM"
    },
    "Id" : "00DbW0000000JDxUAM",
    "Type" : "00D"
  }
} ]
```

**Comment by jongpie** [comment]
Thanks @szymon-halik! I don't see any obvious issue with that query result - I setup the same settings in a test org, and was able to successfully generate a log entry with Nebula Logger's managed package, so something else must be causing the issue. Is this happening in a sandbox (refreshed from your prod org), a scratch org, or some other environment type?

And you mentioned that you already tried the other save methods, but could you try running this anonymous Apex script too? This will skip the platform events altogether, and instead will (try to) write directly to the custom objects `Nebula__Log__c` and `Nebula__LogEntry__c`. You should hopefully either see new records created, or receive a runtime exception (if there's a bug in Nebula Logger).

```apex
Nebula.Logger.error('some error');
Nebula.Logger.saveLog(Nebula.Logger.SaveMethod.SYNCHRONOUS_DML);
```

**Comment by szymon-halik** [comment]
@jongpie We've started to notice this issue this week in newly cloned sandboxes. Our processes of lower envs management looks like that: 
- we create one `base` developer sandbox that is **refreshed** from prod every week and configured accordingly 
- once the configuration procedure is done this sandbox is then later on **cloned** multiple times by admins and devs


We don't use scratch orgs unfortunately yet as we are in the process of decomposing our happy-soup.

I can confirm that above anonymous script produces new records. What is more I tried the following anonymous apex script- I am not sure it that should work/ align with Nebula logger architecture:
```java
Nebula.Logger.error('some error event bus');
Nebula.Logger.saveLog(Nebula.Logger.SaveMethod.EVENT_BUS);
Nebula.Logger.error('some error rest');
Nebula.Logger.saveLog(Nebula.Logger.SaveMethod.REST);
Nebula.Logger.error('some error sync dml');
Nebula.Logger.saveLog(Nebula.Logger.SaveMethod.SYNCHRONOUS_DML);
Nebula.Logger.error('some error queueable');
Nebula.Logger.saveLog(Nebula.Logger.SaveMethod.QUEUEABLE);
```
Results:
- In the impacted(cloned from `base`) sandboxes I receive only log entries from `SYNCHRONOUS_DML` .
- In our full copy sandbox(non impacted so far) with exactly the same config I get entires from `SYNCHRONOUS_DML` and `QUEUEABLE` save methods

**Comment by jongpie** [comment]
@szymon-halik thanks for the extra context about your sandboxes, that's really helpful info to have. Based on the info you've provided, that all sounds fine - I can't think of any reason that your sandbox refresh process would be causing this problem.

It's so very strange that in the impacted environment, the `SYNCHRONOUS_DML` save method works - that save method is typically the more problematic save method (compared to `EVENT_BUS`).
- The `SYNCHRONOUS_DML` save method runs all of the same code as `EVENT_BUS`, it just doesn't call `EventBus.publish()` & instead directly creates the custom object records.
- The fact that you are able to publish platform events in your dummy PE `test__e` indicates that your org has not exceeded any platform event allocation limits, so I'm not sure why Nebula Logger's PE `LogEntryEvent__e` isn't working.

Could you try running this anon Apex script as well? This should help verify that your org can create a `Nebula__LogEntryEvent__e` directly - hopefully you can see this event in the streaming monitor app, and Nebula Logger should also convert it to `Nebula__Log__c` and `Nebula__LogEntry__c` records.

```apex
Nebula__LogEntryEvent__e evt = new Nebula__LogEntryEvent__e(
    Nebula__LoggingLevel__c = System.LoggingLevel.ERROR.name(),
    Nebula__OriginType__c = 'Apex',
    Nebula__Message__c = 'test platform event record',
    Nebula__Timestamp__c = System.now(),
    Nebula__TransactionEntryNumber__c = 1,
    Nebula__TransactionId__c = System.UUID.randomUUID().toString()
);
System.debug(EventBus.publish(evt));
```

**Comment by jongpie** [comment]
@szymon-halik very strange - when I run the anon Apex script, I get essentially the same response as you (my result is below), but within 1-2 seconds, the `Nebula__LogEntryEvent__e` is then successfully converted & saved in the custom objects `Nebula__Log__c` and `Nebula__LogEntry__c`. I'm not really sure what else to try at this point - it might be worth opening a case with Salesforce Support to see if they can do any additional troubleshooting.

I'll continue trying some other ideas on my side, but let me know if you are able to make any progress in your sandboxes.

```json
{
  "id" : "e00xx0000000001AAA",
  "success" : true,
  "errors" : [ {
    "statusCode" : "OPERATION_ENQUEUED",
    "message" : "042b6768-f06f-4139-80ef-56c621e86473",
    "fields" : [ ]
  } ]
}
```

**Comment by james-ef** [comment]
@jongpie I work with @szymon-halik and @Limbo20000.  To add some extra info here.  We create our developer sandboxes by cloning another developer sandbox.  We have just been informed by salesforce that there was a known issue their side for platform event triggers in our cloning scenario.  It was marked as resolved but seems like it is not.

https://issues.salesforce.com/issue/a028c00000iwWAFAA2/

**Comment by szymon-halik** [comment]
Here is a new global issues reference https://issues.salesforce.com/issue/a028c00000yzznTAAQ. It should be done by Summer '24 release

**Comment by jongpie** [comment]
@james-ef, @Limbo20000, and @szymon-halik thanks for the updates! From what I can tell, there doesn't seem to be a workaround for this issue, so hopefully Salesforce will roll out a fix for this soon.

Since this is a platform issue (not an issue with Nebula Logger), I'm going to close this, but please let me know if there's anything else I can do to assist.
